clippy_lints/src/booleans.rs

```diff
@@ -350,8 +350,14 @@ impl SuggestContext<'_, '_, '_> {
                         self.output.push_str(&str);
                     } else {
                         self.output.push('!');
-                        self.output
-                            .push_str(&Sugg::hir_opt(self.cx, terminal)?.maybe_par().to_string());
+                        if terminal.span.from_expansion()
+                            && let Some(orig) = terminal.span.source_callsite().get_source_text(self.cx)
+                        {
+                            self.output.push_str(&orig);
+                        } else {
+                            self.output
+                                .push_str(&Sugg::hir_opt(self.cx, terminal)?.maybe_par().to_string());
+                        }
```

Comment:
You can make this simpler by always requiring the root context:

```suggestion
                        let mut app = Applicability::MachineApplicable;
                        let snip = Sugg::hir_with_context(self.cx, terminal, SyntaxContext::root(), "", &mut app);
                        if app != Applicability::MachineApplicable {
                            return None;
                        }
                        self.output.push_str(&format!("!{}", snip.maybe_par()));
```

(and remove the first `self.output.push('!')`, I cannot include its removal in the diff)

Bonus: you can even make the suggestion better by letting `Sugg` invert the condition by itself. This will change (for the better) the output of one existing test. Just replace the last line by:

```rust
                        self.output.push_str(&(!snip).into_string());
```